ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumptionSpec.scala

```diff
@@ -0,0 +1,143 @@
+        .setConfiguration(Configuration.encode(LedgerReader.DefaultConfiguration))
+        .setSubmissionId("a submission")
+        .setParticipantId("a participant"))
+    .build
+
+  "logEntryToUpdate" should {
+    "throw in case no record time is available from the log entry or input argument" in {
+      assertThrows[Err](
+        logEntryToUpdate(aLogEntryId, aLogEntryWithoutRecordTime, recordTimeForUpdate = None))
+    }
+
+    "use log entry's record time instead of one provided as input" in {
+      val actual :: Nil = logEntryToUpdate(
+        aLogEntryId,
+        aLogEntryWithRecordTime,
+        recordTimeForUpdate = Some(aRecordTimeForUpdate))
+
+      actual.recordTime shouldBe aRecordTimeFromLogEntry
+    }
+
+    "use record time from log entry if not provided as input" in {
+      val actual :: Nil =
+        logEntryToUpdate(aLogEntryId, aLogEntryWithRecordTime, recordTimeForUpdate = None)
+
+      actual.recordTime shouldBe Timestamp.assertFromInstant(Instant.ofEpochSecond(100))
+    }
+  }
+
+  "outOfTimeBoundsEntryToUpdate" should {
+    "not generate an update for a deduplicated transaction entry" in {
+      val inputEntry = DamlOutOfTimeBoundsEntry.newBuilder
+        .setEntry(aLogEntryWithTransactionRejectionEntry)
+        .setDuplicateUntil(
+          Conversions.buildTimestamp(Timestamp.assertFromInstant(aRecordTimeForUpdateInstant)))
+        .build
+
+      outOfTimeBoundsEntryToUpdate(aRecordTimeForUpdate, inputEntry) shouldBe None
+    }
+
+    "generate a rejection entry for a transaction if record time is too late" in {
+      val inputEntry = DamlOutOfTimeBoundsEntry.newBuilder
+        .setEntry(aLogEntryWithTransactionRejectionEntry)
+        .setTooLateFrom(Conversions.buildTimestamp(
+          Timestamp.assertFromInstant(aRecordTimeForUpdateInstant.minusMillis(1))))
+        .build
+
+      val expectedUpdate = Update.CommandRejected(
+        recordTime = aRecordTimeForUpdate,
+        submitterInfo = Conversions.parseSubmitterInfo(someSubmitterInfo),
+        reason = RejectionReason.InvalidLedgerTime("Ledger time outside of valid range")
+      )
+      outOfTimeBoundsEntryToUpdate(aRecordTimeForUpdate, inputEntry) shouldBe Some(expectedUpdate)
+    }
+
+    "generate a rejection entry for a transaction if record time is too early" in {
+      val inputEntry = DamlOutOfTimeBoundsEntry.newBuilder
+        .setEntry(aLogEntryWithTransactionRejectionEntry)
+        .setTooEarlyUntil(Conversions.buildTimestamp(
+          Timestamp.assertFromInstant(aRecordTimeForUpdateInstant.plusMillis(1))))
+        .build
+
+      val expectedUpdate = Update.CommandRejected(
+        recordTime = aRecordTimeForUpdate,
+        submitterInfo = Conversions.parseSubmitterInfo(someSubmitterInfo),
+        reason = RejectionReason.InvalidLedgerTime("Ledger time outside of valid range")
+      )
+      outOfTimeBoundsEntryToUpdate(aRecordTimeForUpdate, inputEntry) shouldBe Some(expectedUpdate)
+    }
+
```

Comment:
A test case seems to be missing for the case when both bounds are present. Also, see my comment above RE the messages for the single bound.

Reply:
Added test case for that.